pkg/controller/smart-clone-controller.go

```diff
@@ -0,0 +1,329 @@
+	expectations "kubevirt.io/containerized-data-importer/pkg/expectations"
+	csiclientset "kubevirt.io/containerized-data-importer/pkg/snapshot-client/clientset/versioned"
+	snapshotsinformers "kubevirt.io/containerized-data-importer/pkg/snapshot-client/informers/externalversions/volumesnapshot/v1alpha1"
+	snapshotslisters "kubevirt.io/containerized-data-importer/pkg/snapshot-client/listers/volumesnapshot/v1alpha1"
+)
+
+const (
+	//AnnSmartCloneRequest sets our expected annotation for a CloneRequest
+	AnnSmartCloneRequest = "k8s.io/SmartCloneRequest"
+)
+
+// SmartCloneController represents the CDI SmartClone Controller
+type SmartCloneController struct {
+	Controller
+	cdiClientSet clientset.Interface
+	csiClientSet csiclientset.Interface
+
+	snapshotInformer  cache.SharedIndexInformer
+	snapshotsLister   snapshotslisters.VolumeSnapshotLister
+	dataVolumeLister  listers.DataVolumeLister
+	dataVolumesSynced cache.InformerSynced
+	snapshotsSynced   cache.InformerSynced
+	podExpectations   *expectations.UIDTrackingControllerExpectations
+}
+
+// NewSmartCloneController sets up a Smart Clone Controller, and returns a pointer to
+// to the newly created Controller
+func NewSmartCloneController(client kubernetes.Interface,
+	cdiClientSet clientset.Interface,
+	csiClientSet csiclientset.Interface,
+	pvcInformer coreinformers.PersistentVolumeClaimInformer,
+	podInformer coreinformers.PodInformer,
+	snapshotInformer snapshotsinformers.VolumeSnapshotInformer,
+	dataVolumeInformer informers.DataVolumeInformer,
+	image string,
+	pullPolicy string,
+	verbose string) *SmartCloneController {
+	c := &SmartCloneController{
+		Controller:        *NewController(client, pvcInformer, podInformer, image, pullPolicy, verbose),
+		cdiClientSet:      cdiClientSet,
+		csiClientSet:      csiClientSet,
+		snapshotsLister:   snapshotInformer.Lister(),
+		snapshotInformer:  snapshotInformer.Informer(),
+		dataVolumeLister:  dataVolumeInformer.Lister(),
+		dataVolumesSynced: dataVolumeInformer.Informer().HasSynced,
+		snapshotsSynced:   snapshotInformer.Informer().HasSynced,
+		podExpectations:   expectations.NewUIDTrackingControllerExpectations(expectations.NewControllerExpectations()),
+	}
+
+	// Set up an event handler for when VolumeSnapshot resources change
+	snapshotInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: c.enqueueVolumeSnapshot,
+		UpdateFunc: func(old, new interface{}) {
+			c.enqueueVolumeSnapshot(new)
+		},
+		DeleteFunc: c.enqueueVolumeSnapshot,
+	})
+
+	// Set up an event handler for when PVC resources change
+	// handleObject function ensures we filter PVCs not created by this controller
+	pvcInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: c.enqueuePvc,
+		UpdateFunc: func(old, new interface{}) {
+			c.enqueuePvc(new)
+		},
+		DeleteFunc: c.enqueuePvc,
+	})
+
+	return c
+}
+
+func (c *SmartCloneController) initializeExpectations(pvcKey string) error {
```

Comment:
Do we need these two? I don't think pods are created in this controller.

Reply:
Renamed